### PR TITLE
soc: arm: st_stm32: stm32l4: power.c standby shutdown mode

### PR DESCRIPTION
**Pull request by Desvauxm-st**
STM32L4x power management of Standby mode
and shutdown mode

### CODE REVIEW

**Review by FRASTM — CHANGES_REQUESTED** [review]
* go with 2 commits, at least ; one for sample one  for soc 
* fix CI warnings
* detail the  correlation with sleep "K_TICKS_FOREVER" if any

File: soc/arm/st_stm32/stm32l4/power.c [hunk 1]
```diff
@@ -20,89 +20,139 @@
...
-			substate_id);
+		LOG_DBG("Unsupported power state substate-id %u", substate_id);
+		break;
+	}
+}
+
+void set_mode_standby(void)
+{
+	/* Select standby mode */
+	printk(" mode_standby done\n\n\n");
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:81]: No printk. Replace by LOG_BDG

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	uint32_t cause;
+
+	hwinfo_get_reset_cause(&cause);
+	hwinfo_clear_reset_cause();
+
+	if ((LL_PWR_IsActiveFlag_SB() == true) && (cause == 0))	{
+		LL_PWR_ClearFlag_SB();
+		LL_PWR_ClearFlag_WU();
+		printk(" reset from Standby mode exit\n");
+		printk(" reset from Standby mode exit\n");
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:68]:
```suggestion
		printk("reset from Standby mode exit\n");
```
Apply everywhere

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	if (cause == (RESET_PIN | RESET_BROWNOUT)) {
+		LL_PWR_ClearFlag_WU();
+		printk(" reset from shutdown mode exit or power up\n");
+		printk(" reset from shutdown mode exit or power up\n");
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:74]: Why duplicating all printk ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/pm/pm.h>
+
+
+#define SLEEP_TIME_MS   300
+
+
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:18]: #if !defined(CONFIG_SOC_SERIES_STM32L4X)
#error Not implemented for other series
#endif

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 5]
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:2]: Fix copyright

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+#include <stm32_ll_pwr.h>
+#define SW0_NODE	DT_ALIAS(sw0)
+#if !DT_NODE_HAS_STATUS(SW0_NODE, okay)
+#error "Unsupported board: sw0 devicetree alias is not defined"
+#endif
+
+#if !defined(CONFIG_SOC_SERIES_STM32L4X)
+#error Not implemented for other series
+#endif  /* CONFIG_SOC_SERIES_STM32L4X */
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:27]: Move this earlier

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+static const struct gpio_dt_spec button = GPIO_DT_SPEC_GET_OR(SW0_NODE, gpios,
+							      {0});
+
+
+static const struct gpio_dt_spec led =
+	GPIO_DT_SPEC_GET(DT_ALIAS(led0), gpios);
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:34]: Use consistent indentation

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	/* Enable pin pull up configurations and wakeup pins */
+	LL_PWR_EnablePUPDCfg();
+	LL_PWR_EnableWakeUpPin(LL_PWR_WAKEUP_PIN2);
+	/* Clear wakeup flags */
+	LL_PWR_ClearFlag_WU();
+}
+
+
+
+
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:52]: Remove superflous lines.
Apply everywhere

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+void main(void)
+{
+
+
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:56]: Superfluous lines

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+	printk(" Device ready :%s\n\n\n", CONFIG_BOARD);
+
+	printk("press and hold the user button to enter to Shutdown Mode when LED2 is OFF\n\n");
+	printk("press and hold the user button to enter to Standby Mode when LED2 is ON\n\n");
+
+	led_is_on = true;
+	while (true) {
+		int i;
+
+		for (i = 1 ; i < 10 ; i++) {
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:103]: Is the for loop really required ?

File: soc/arm/st_stm32/stm32l4/power.c [hunk 11]
```diff
@@ -20,89 +20,140 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
 
+
+
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:24]: Remove these lines

File: soc/arm/st_stm32/stm32l4/power.c [hunk 12]
```diff
@@ -20,89 +20,140 @@
...
 #if STM32_SYSCLK_SRC_MSI
 #define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_MSI
 #else
 #define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_HSI
 #endif
 
-/* Invoke Low Power/System Off specific Tasks */
-__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+
+
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:33]: remove these lines

File: soc/arm/st_stm32/stm32l4/power.c [hunk 13]
```diff
@@ -20,89 +20,140 @@
...
+void enter_low_power_mode(void)
 {
-	if (state != PM_STATE_SUSPEND_TO_IDLE) {
-		LOG_DBG("Unsupported power state %u", state);
-		return;
+
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:36]: remove this line

File: soc/arm/st_stm32/stm32l4/power.c [hunk 14]
```diff
@@ -20,89 +20,140 @@
...
+	/* Configure CPU core */
+	/* Enable CPU deep sleep mode */
+#ifndef CONFIG_DEBUG
+	LL_LPM_EnableDeepSleep();
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:40]: Is this required before shutdown/standby ?

File: soc/arm/st_stm32/stm32l4/power.c [hunk 15]
```diff
@@ -20,89 +20,140 @@
...
 #ifdef PWR_CR1_RRSTP
 		LL_PWR_DisableSRAM3Retention();
 #endif /* PWR_CR1_RRSTP */
 		/* enter STOP2 mode */
 		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP2);
+		break;
+	default:
+		LOG_DBG("Unsupported power state substate-id %u", substate_id);
+		break;
+	}
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:76]: k_cpu_idle();

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 10 above)
Comment (erwango) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:103]: Please use GPIO interrupt if reactivity required

File: soc/arm/st_stm32/stm32l4/power.c (hunk 14 above)
Comment (Desvauxm-st) [reply soc/arm/st_stm32/stm32l4/power.c:40]: yes

File: soc/arm/st_stm32/stm32l4/power.c (hunk 15 above)
Comment (Desvauxm-st) [reply soc/arm/st_stm32/stm32l4/power.c:76]: k_cpu_idle(); move in pm_state_set() only for PM_STATE_SUSPEND_TO_IDLE

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+{
+	k_sem_give(&button_sem);
+}
+
+void thread_Shutdown_Standby_mode(void)
+{
+	int sleep_ms;
+	sleep_ms = 300;
+	k_sem_init(&button_sem, 0, 1);
+	while (true) {
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:68]: I don't think while loop is required

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+			config_wakeup_features();
+			pm_state_force(0u, &(struct pm_state_info) {PM_STATE_SOFT_OFF, 0, 0});
+
+			k_msleep(200);
+			while (1) {
+			/* stay in Shutdown mode until wakeup line activated */
+				k_msleep(200);
+			}
+		}
+		if (led_is_on == true) {
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:84]: else

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+	int sleep_ms;
+	sleep_ms = 300;
+	k_sem_init(&button_sem, 0, 1);
+	while (true) {
+		k_sem_take(&button_sem, K_FOREVER);
+		if (led_is_on == false) {
+			gpio_pin_configure(led.port, led.pin, GPIO_DISCONNECTED);
+			printk("user_button pressed\n");
+			printk("enter into Shutdown Mode\n");
+			printk("release the user button to exit from Shutdown Mode\n\n");
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:74]: Let's factorize when possible

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+			printk("enter into Standby Mode\n");
+			printk("release the user button to exit from Standby Mode\n\n");
+			config_wakeup_features();
+			pm_state_force(0u, &(struct pm_state_info) {PM_STATE_STANDBY, 0, 0});
+
+			k_msleep(200);
+			while (1) {
+			/* stay in Standby mode until wakeup line activated */
+				k_msleep(200);
+			}
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:96]: This shouldn't be required. Can you have a new try w/o ?

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+
+	k_sem_init(&button_sem, 0, 1);
+	while (true) {
+		k_sem_take(&button_sem, K_FOREVER);
+		if (led_is_on == false) {
+			gpio_pin_configure(led.port, led.pin, GPIO_DISCONNECTED);
+			printk("user_button pressed\n");
+			printk("enter into Shutdown Mode\n");
+			printk("release the user button to exit from Shutdown Mode\n\n");
+			config_wakeup_features();
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:75]: This doesn't have to be done after button press.
Can you move it earlier ?

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+
+void main(void)
+{
+	int ret;
+	uint32_t cause;
+
+	hwinfo_get_reset_cause(&cause);
+	hwinfo_clear_reset_cause();
+
+	if ((LL_PWR_IsActiveFlag_SB() == true) && (cause == 0))	{
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:113]: It would be great to have LL_PWR_IsActiveFlag_SB detection as part of `hwinfo_get_reset_cause`.
Can you check if possible to do this ?
(This would be another PR)

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+		LL_PWR_ClearFlag_SB();
+		LL_PWR_ClearFlag_WU();
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:115]: Should it be done in `hwinfo_clear_reset_cause();` ?
(to be done in the hwinfo PR)

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+	if (ret != 0) {
+		printk("Error %d: failed to configure interrupt on %s pin %d\n",
+			ret, button.port->name, button.pin);
+		return;
+	}
+
+	gpio_init_callback(&button_cb_data, button_pressed, BIT(button.pin));
+	gpio_add_callback(button.port, &button_cb_data);
+
+	printk(" Device ready :%s\n\n\n", CONFIG_BOARD);
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:155]:
```suggestion
	printk("Device ready: %s\n\n\n", CONFIG_BOARD);
```

**Review by erwango — CHANGES_REQUESTED** [review]
Some more changes requested, but globally looks good.

Regarding the app, can you work on readability?
For instance, following would be easier to read:
```
*** Booting Zephyr OS build zephyr-v3.3.0-619-g693bf8e46f4b ***   
                                                                  
Reset cause: Shutdown Mode                                            
                                                                  
Device ready: b_l4s5i_iot01a                                      
                                                                  
Press and hold the user button:                                   
  when LED2 is OFF to enter to Shutdown Mode                      
  when LED2 is ON to enter to Standby Mode                        
                                                                  
User button pressed: 
Standby Mode requested         
                                 
Release the user button to exit from Standby Mode                 
                                                                  
*** Booting Zephyr OS build zephyr-v3.3.0-619-g693bf8e46f4b ***   

```

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 23 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:155]: ok

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 16 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:68]: OK

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 17 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:84]: ok

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 19 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:96]: Yes right

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 20 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:75]: yes

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 21 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:113]: Yes I will try to check if it possible to do that

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 22 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:115]: Yes I will try to check if it possible to do that

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/sample.yaml [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+  name: STM32 GPIO Power Management
+tests:
+  sample.boards.stm32.power_mgmt.blinky_standby_shutdown_mode:
+    platform_allow: nucleo_l476rg
+    tags: LED power
+    harness: console
+    harness_config:
+      type: one_line
+      regex:
+        - "Device ready"
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/sample.yaml:11]: This will fail in CI. 
First message when running twister is likely `Reset cause: Reset pin`

**Review by erwango — CHANGES_REQUESTED** [review]
Few more comments.

Last: Please split in 3 commits:
- Update power.c to factorise stop states
- Update power.c to add standby and shutdown
- Add the sample

Also rename the sample to: standby_shutdown

Thanks

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+		printk("Release the user button to exit from Shutdown Mode\n\n");
+		pm_state_force(0u, &(struct pm_state_info) {PM_STATE_SOFT_OFF, 0, 0});
+		/* stay in Shutdown mode until wakeup line activated */
+	} else {
+		printk("Standby Mode requested\n");
+		printk("Release the user button to exit from Standby Mode\n\n");
+		pm_state_force(0u, &(struct pm_state_info) {PM_STATE_STANDBY, 0, 0});
+		/* stay in Standby mode until wakeup line activated */
+	}
+	k_msleep(200);
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:83]: Is that strictly required ?

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c [hunk 26]
```diff
@@ -0,0 +1,153 @@
...
+
+#if !defined(CONFIG_SOC_SERIES_STM32L4X)
+#error Not implemented for other series
+#endif  /* CONFIG_SOC_SERIES_STM32L4X */
+
+#define STACKSIZE 1024
+#define PRIORITY 7
+#define SLEEP_TIME_MS   3000
+
+#include <stm32_ll_pwr.h>
```
[the remark this answers is not in the document]
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c]: yes

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/sample.yaml (hunk 24 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/sample.yaml]: ok

File: samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c (hunk 25 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:83]: no works without k_msleep

File: soc/arm/st_stm32/stm32l4/power.c [hunk 27]
```diff
@@ -27,82 +27,93 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
...
+}
+
+/* Invoke Low Power/System Off specific Tasks */
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	switch (state) {
+	case PM_STATE_SUSPEND_TO_IDLE:
+		set_mode_stop(substate_id);
+	/* Set SLEEPDEEP bit of Cortex System Control Register */
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:95]: Fix comment indentation

File: soc/arm/st_stm32/stm32l4/power.c [hunk 28]
```diff
@@ -54,6 +70,22 @@ void set_mode_stop(uint8_t substate_id)
 	}
 }
 
+void set_mode_standby(void)
+{
+	/* Select standby mode */
+	LOG_DBG("mode_standby done\n\n\n");
```
[the remark this answers is not in the document]
Comment (erwango) [reply soc/arm/st_stm32/stm32l4/power.c:76]: Otherwise, remove these logs.

File: samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c [hunk 29]
```diff
@@ -0,0 +1,152 @@
...
+		    uint32_t pins)
+{
+	k_sem_give(&button_sem);
+}
+
+void thread_Shutdown_Standby_mode(void)
+{
+	int sleep_ms;
+
+	sleep_ms = 300;
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:66]: Not required anymore

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32l4/power.c [hunk 30]
```diff
@@ -27,82 +27,93 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
...
+/* Invoke Low Power/System Off specific Tasks */
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	switch (state) {
+	case PM_STATE_SUSPEND_TO_IDLE:
+		set_mode_stop(substate_id);
+	/* Set SLEEPDEEP bit of Cortex System Control Register */
 		LL_LPM_EnableDeepSleep();
-		/* enter SLEEP mode : WFE or WFI */
+	/* Select mode entry : WFE or WFI and enter the CPU selected mode */
```
[the remark this answers is not in the document]
Comment (Desvauxm-st) [reply soc/arm/st_stm32/stm32l4/power.c:97]: ok

File: samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c (hunk 29 above)
Comment (Desvauxm-st) [reply samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:66]: ok

File: soc/arm/st_stm32/stm32l4/power.c [hunk 31]
```diff
@@ -27,6 +27,22 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
 #define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_HSI
 #endif
 
+void enter_low_power_mode(void)
```
[the remark this answers is not in the document]
Comment (Desvauxm-st) [reply soc/arm/st_stm32/stm32l4/power.c:30]: ok

File: soc/arm/st_stm32/stm32l4/power.c [hunk 32]
```diff
@@ -27,6 +27,22 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
...
+{
+	/* Configure CPU core */
+	/* Enable CPU deep sleep mode */
+#ifndef CONFIG_DEBUG
+	LL_LPM_EnableDeepSleep();
```
[the remark this answers is not in the document]
Comment (Desvauxm-st) [reply soc/arm/st_stm32/stm32l4/power.c:35]: ok

File: samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c [hunk 33]
```diff
@@ -0,0 +1,155 @@
...
+	k_sem_take(&button_sem, K_FOREVER);
+	gpio_pin_configure(led.port, led.pin, GPIO_DISCONNECTED);
+	printk("User button pressed\n");
+	config_wakeup_features();
+	if (led_is_on == false) {
+		printk("Shutdown Mode requested\n");
+		printk("Release the user button to exit from Shutdown Mode\n\n");
+#ifdef CONFIG_LOG
+		k_msleep(2000);
+#endif /* CONFIG_LOG */
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:74]: 2 sec seems quite long for user. Likely user would release the button before that.
Also 2 sec seems large to print a 30 char line, what about 200 ms  ?

File: samples/boards/stm32/power_mgmt/standby_shutdown/sample.yaml [hunk 34]
```diff
@@ -0,0 +1,19 @@
...
+tests:
+  sample.boards.stm32.power_mgmt.standby_shutdown:
+    platform_allow: nucleo_l476rg
+    tags: LED power
+    harness: console
+    harness_config:
+      type: multi_line
+      regex:
+        - "Reset cause: Reset pin"
+        - "Device ready: nucleo_l476rg"
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/standby_shutdown/sample.yaml:12]: - "Device ready: .*"

File: samples/boards/stm32/power_mgmt/standby_shutdown/sample.yaml [hunk 35]
```diff
@@ -0,0 +1,19 @@
+sample:
+  name: STM32 GPIO Power Management
+tests:
+  sample.boards.stm32.power_mgmt.standby_shutdown:
+    platform_allow: nucleo_l476rg
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/standby_shutdown/sample.yaml:5]: It would be nice to get this working on disco_l475_iot1 which is the L4 target used on bench.
Can we instead filter on CONFIG_SOC_SERIES_STM32L4X ?
(I guess the only condition is to know if config_wakeup_features is also valid for most other L4 targets)

File: samples/boards/stm32/power_mgmt/standby_shutdown/README.rst [hunk 36]
```diff
@@ -0,0 +1,49 @@
...
+when LED2 is ON to enter to Standby Mode
+release the user button to exit from shutdown mode or from shutdown mode.
+
+
+
+PM configurations
+*****************
+
+By default, :kconfig:option:`CONFIG_PM_DEVICE` and `CONFIG_HWINFO` are enabled.
+for another board than nucleo_L476RG please adjust wakeup pin into config_wakeup_features()
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/standby_shutdown/README.rst:49]: Move this in Requirements

File: samples/boards/stm32/power_mgmt/standby_shutdown/README.rst [hunk 37]
```diff
@@ -0,0 +1,49 @@
...
+   :goals: build flash
+   :compact:
+
+After flashing, the LED starts to blink.
+Press and hold the user button:
+when LED2 is OFF to enter to Shutdown Mode
+when LED2 is ON to enter to Standby Mode
+release the user button to exit from shutdown mode or from shutdown mode.
+
+
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/standby_shutdown/README.rst:43]: Remove this line

**Review by erwango — DISMISSED** [review]

File: samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c [hunk 38]
```diff
@@ -0,0 +1,155 @@
...
+	k_sem_init(&button_sem, 0, 1);
+	k_sem_take(&button_sem, K_FOREVER);
+	gpio_pin_configure(led.port, led.pin, GPIO_DISCONNECTED);
+	printk("User button pressed\n");
+	config_wakeup_features();
+	if (led_is_on == false) {
+		printk("Shutdown Mode requested\n");
+		printk("Release the user button to exit from Shutdown Mode\n\n");
+#ifdef CONFIG_LOG
+		k_msleep(2000);
```
Comment (FRASTM) [inline samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:73]: why sleeping for 2s when CONFIG_LOG ?

**Review by FRASTM — COMMENTED** [review]
LGTM

File: samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c [hunk 39]
```diff
@@ -0,0 +1,155 @@
...
+	LL_PWR_ClearFlag_WU();
+}
+
+void button_pressed(const struct device *dev, struct gpio_callback *cb,
+		    uint32_t pins)
+{
+	k_sem_give(&button_sem);
+}
+
+void thread_Shutdown_Standby_mode(void)
```
Comment (tagunil) [inline samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:62]: Shouldn't that be lowercase?

File: soc/arm/st_stm32/stm32l4/power.c [hunk 40]
```diff
@@ -27,82 +27,124 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
...
-	if (state != PM_STATE_SUSPEND_TO_IDLE) {
-		LOG_DBG("Unsupported power state %u", state);
-		return;
+	/* Configure CPU core */
+	/* Enable CPU deep sleep mode */
+	LL_LPM_EnableDeepSleep();
+	LL_DBGMCU_DisableDBGStandbyMode();
+	/* Enter ultra_low-power mode */
+
+	for (;;) {
```
Comment (tagunil) [inline soc/arm/st_stm32/stm32l4/power.c:38]: Is that actually better than just calling k_cpu_idle()?

Comment (tagunil) [reply soc/arm/st_stm32/stm32l4/power.c:38]: Sorry, I haven't been clear enough. The question was not about the loop, but about adding a separate wfi path instead of going through k_cpu_idle() which also does wfi.

Comment (Desvauxm-st) [reply soc/arm/st_stm32/stm32l4/power.c:38]: OK I'll take a look and test if this change works.

Comment (tagunil) [reply soc/arm/st_stm32/stm32l4/power.c:38]: I mean, why not just adapt the approach used in stm32u5/power.c? I think the code could be almost identical.

**Comment by tagunil** [comment]
@erwango @Desvauxm-st @FRASTM what do you think about using the STM32U5 power management implementation as a base for the shared one? I think the implementation could be unified at least for L0/L1/L4/L5/G0/G4/U5 using some ifdefs. Not sure about dual-core ones.

**Comment by erwango** [comment]
> @erwango @Desvauxm-st @FRASTM what do you think about using the STM32U5 power management implementation as a base for the shared one? I think the implementation could be unified at least for L0/L1/L4/L5/G0/G4/U5 using some ifdefs. Not sure about the dual-core ones.

I'm not strongly opposed to it. This would bring some benefits indeed. My main request would be to keep it the less cluttered possible and easy to opt-out (I have in mind some upcoming series which would require real divergence from the main scheme). Are you planning to work on that ?

**Comment by tagunil** [comment]
> > @erwango @Desvauxm-st @FRASTM what do you think about using the STM32U5 power management implementation as a base for the shared one? I think the implementation could be unified at least for L0/L1/L4/L5/G0/G4/U5 using some ifdefs. Not sure about the dual-core ones.
> 
> I'm not strongly opposed to it. This would bring some benefits indeed. My main request would be to keep it the less cluttered possible and easy to opt-out (I have in mind some upcoming series which would require real divergence from the main scheme). Are you planning to work on that ?

I do. One of my new projects requires consistent power management support across several series, so I think common implementation could be useful for that.

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]
